curriculum/challenges/english/25-front-end-development/lab-permutation-generator/66fe4f33a2cc9b33f4d5cd9b.md

```diff
@@ -0,0 +1,159 @@
+---
+id: 66fe4f33a2cc9b33f4d5cd9b
+title: Build a Permutation Generator
+challengeType: 14
+dashedName: build-a-permutation-generator
+---
+
+# --description--
+
+In this lab, you will build a permutation generator that will take a string and return all possible permutations of the characters in the string. For example, the possible permutations of the string `'cat'` are `'cat'`, `'cta'`, `'act'`, `'atc'`, `'tac'`, and `'tca'`.
+
+**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 
+
+**User Stories:**
+
+1. You should create a function named `permuteString`.
+2. The `permuteString` function should take three parameters- a string, a prefix value and an empty array for storing and returning the results. The prefix value would accumulate characters to form a permutation.
+3. Inside the function, you should check if the length of the passed string is `0`. If it is, push the current prefix to the results and return the results.
+4. Iterate over each character in the input string and for each iteration, remove the current character from the string and call the `permuteString` function recursively with updated arguments to build the remaining permutations.
+5. You should return the final results array.
+6. You should ensure that the permutations are unique by removing duplicates.
+
+# --hints--
+
+You should have a function `permuteString`.
+
+```js
+assert.isFunction(permuteString);
+```
+
+`permuteString('far')` should return `[ 'far', 'fra', 'afr', 'arf', 'rfa', 'raf' ]`.
+
+```js
+assert.sameMembers(permuteString('far'), [ 'far', 'fra', 'afr', 'arf', 'rfa', 'raf' ]);
+```
+
+`permuteString('fcc')` should return `[ 'fcc', 'cfc', 'ccf' ]`.
+
+```js
+assert.sameMembers(permuteString('fcc'), [ 'fcc', 'cfc', 'ccf' ]);
+```
+
+`permuteString('p')` should return `[ 'p' ]`.
+
+```js
+assert.deepStrictEqual(permuteString('p'), ['p']);
+```
+
+`permuteString('')` should return `['']`.
+
+```js
+let emptyArr= permuteString('');
+
+// 1 because the empty string is being pushed and is a permutation of itself
+assert.lengthOf(emptyArr, 1); 
+
+```
+
+`permuteString('101')` should return `[ '101', '110', '011', '110', '101' ]`.
```

Comment:
This test is asking for duplicates - `101` and `110` are in there twice. The function should not return duplicates, correct? It also means that the solution that is passing this test isn't removing duplicates properly.